File: Data Structures & Algorithms/products-of-array-discluding-self/submission-14.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int prod = 1, zeroCount = 0;
        foreach (int num in nums) {
            if (num != 0) {
                prod *= num;
            } else {
                zeroCount++;
            }
        }

        if (zeroCount > 1) {
            return new int[nums.Length]; 
        }

        int[] res = new int[nums.Length];
        for (int i = 0; i < nums.Length; i++) {
            if (zeroCount > 0) {
                res[i] = (nums[i] == 0) ? prod : 0;
            } else {
                res[i] = prod / nums[i];
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-14.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxProduct(nums) {
        let A = [];
        let cur = [];
        let res = -Infinity;

        nums.forEach(num => {
            res = Math.max(res, num);
            if (num === 0) {
                if (cur.length) A.push(cur);
                cur = [];
            } else {
                cur.push(num);
            }
        });
        if (cur.length) A.push(cur);

        A.forEach(sub => {
            let negs = 0;
            sub.forEach(i => {
                if (i < 0) negs++;
            });

            let prod = 1;
            let need = (negs % 2 === 0) ? negs : (negs - 1);
            negs = 0;
            for (let i = 0, j = 0; i < sub.length; i++) {
                prod *= sub[i];
                if (sub[i] < 0) {
                    negs++;
                    while (negs > need) {
                        prod /= sub[j];
                        if (sub[j] < 0) negs--;
                        j++;
                    }
                }
                if (j <= i) res = Math.max(res, prod);
            }
        });

        return res;
    }
}